# Exercise

1. Use famous iris flower dataset from sklearn.datasets to predict flower species using SVM classifier.
2. Measure prediction score by using different Kernal Function
3. Find the best classifer and score for the Kernal Function

# Load the iris flower dataset

In [151]:
# Write your Code Here
# Importing the libraries and dataset
# Importing the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC

In [152]:
# Write your Code Here
# Importing the dataset
iris = pd.read_csv('/content/iris.csv')
iris

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

## Understand the dataset
1. iris.DESCR > Complete description of dataset
2. iris.data > Data to learn. Each training set is 4 digit array of features. Total 150 training sets
3. iris.feature_names > Array of all 4 feature ['sepal length (cm)','sepal width cm)','petal length (cm)','petal width (cm)']
4. iris.filename > CSV file name
5. iris.target > The classification label. For every training set there is one classification label(0,1,2). Here 0 for setosa, 1 for versicolor and 2 for virginica
6. iris.target_names > the meaning of the features. Its a array >> ['setosa', 'versicolor', 'virginica']
From above details its clear that our data is 'iris.data' and labels are 'iris.target'

Each training example in iris.data is a array of iris.feature_names lets create a dataframe

# Set the Input and Target Column

In [153]:
#Write your Code Here
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
iris['Species'] = le.fit_transform(iris['Species'])

In [154]:
X = iris.iloc[:, [1,4]].values
Y = iris.iloc[:, 5].values

# Display the shape of the dataset with head and tail portion

In [155]:
#Write your Code Here
print(iris.shape)
print('Head\n', iris.head(5))
print('Tail\n', iris.tail(5))

(150, 6)
Head
    Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0   1            5.1           3.5            1.4           0.2        0
1   2            4.9           3.0            1.4           0.2        0
2   3            4.7           3.2            1.3           0.2        0
3   4            4.6           3.1            1.5           0.2        0
4   5            5.0           3.6            1.4           0.2        0
Tail
       Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
145  146            6.7           3.0            5.2           2.3        2
146  147            6.3           2.5            5.0           1.9        2
147  148            6.5           3.0            5.2           2.0        2
148  149            6.2           3.4            5.4           2.3        2
149  150            5.9           3.0            5.1           1.8        2


# Preprocessing missing values

In [156]:
#Write your Code Here
#Write your Code Here
print(iris.isnull().sum())

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64


In [157]:
#Write your Code Here
iris.dtypes

Id                 int64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species            int64
dtype: object

# Preprocessing Duplicate values

In [158]:
#Write your Code Here
print(iris.duplicated())
print(iris.duplicated().sum())

0      False
1      False
2      False
3      False
4      False
       ...  
145    False
146    False
147    False
148    False
149    False
Length: 150, dtype: bool
0


#  Split the dataset into the Training set and Test set

In [159]:
X_Train, X_Test, Y_Train, Y_Test = train_test_split(X, Y, test_size = 0.25, random_state = 0)

# Feature scaling

In [160]:
#Write your Code Here
sc_X = StandardScaler()
X_Train = sc_X.fit_transform(X_Train)
X_Test = sc_X.transform(X_Test)

#   Fitting the classifier into the Training set

In [161]:
#Write your Code Here
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_Train, Y_Train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

#   Predicting the test set results

In [162]:
#Write your Code Here
Y_Pred = classifier.predict(X_Test)

#   Display the confusion matrix

In [163]:
#Write your Code Here
from sklearn.metrics import classification_report, confusion_matrix
cm = confusion_matrix(Y_Test, Y_Pred)

In [164]:
print(confusion_matrix(Y_Test, Y_Pred))

[[13  0  0]
 [ 0 16  0]
 [ 0  1  8]]


#   Display the Classification Report

In [165]:
print(classification_report(Y_Test, Y_Pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        13
           1       0.94      1.00      0.97        16
           2       1.00      0.89      0.94         9

    accuracy                           0.97        38
   macro avg       0.98      0.96      0.97        38
weighted avg       0.98      0.97      0.97        38



Try by using different Kernal function

LINEAR KERNEL

In [166]:
#Write your Code Here
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_Train, Y_Train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

In [167]:
Y_Pred = classifier.predict(X_Test)

In [168]:
classifier.score(X_Train,Y_Train)

0.9553571428571429

In [169]:
classifier.score(X_Test,Y_Test)

0.9736842105263158

In [170]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_Test, Y_Pred)

0.9736842105263158

In [171]:
from sklearn.metrics import classification_report, confusion_matrix
cm = confusion_matrix(Y_Test, Y_Pred)

In [172]:
print(confusion_matrix(Y_Test, Y_Pred))

[[13  0  0]
 [ 0 16  0]
 [ 0  1  8]]


In [173]:
print(classification_report(Y_Test, Y_Pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        13
           1       0.94      1.00      0.97        16
           2       1.00      0.89      0.94         9

    accuracy                           0.97        38
   macro avg       0.98      0.96      0.97        38
weighted avg       0.98      0.97      0.97        38



SIGMOID KERNEL

In [174]:
#Write your Code Here
from sklearn.svm import SVC
classifier = SVC(kernel = 'sigmoid', random_state = 0)
classifier.fit(X_Train, Y_Train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='sigmoid',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

In [175]:
Y_Pred = classifier.predict(X_Test)

In [176]:
classifier.score(X_Train,Y_Train)

0.7857142857142857

In [177]:
classifier.score(X_Test,Y_Test)

0.8157894736842105

In [178]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_Test, Y_Pred)

0.8157894736842105

In [179]:
from sklearn.metrics import classification_report, confusion_matrix
cm = confusion_matrix(Y_Test, Y_Pred)

In [180]:
print(confusion_matrix(Y_Test, Y_Pred))

[[13  0  0]
 [ 1 10  5]
 [ 0  1  8]]


In [181]:
print(classification_report(Y_Test, Y_Pred))

              precision    recall  f1-score   support

           0       0.93      1.00      0.96        13
           1       0.91      0.62      0.74        16
           2       0.62      0.89      0.73         9

    accuracy                           0.82        38
   macro avg       0.82      0.84      0.81        38
weighted avg       0.85      0.82      0.81        38



POLYNOMIAL KERNEL

In [182]:
#Write your Code Here
from sklearn.svm import SVC
classifier = SVC(kernel = 'poly', random_state = 0)
classifier.fit(X_Train, Y_Train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='poly',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

In [183]:
Y_Pred = classifier.predict(X_Test)

In [184]:
classifier.score(X_Train,Y_Train)

0.9017857142857143

In [185]:
classifier.score(X_Test,Y_Test)

0.868421052631579

In [186]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_Test, Y_Pred)

0.868421052631579

In [187]:
from sklearn.metrics import classification_report, confusion_matrix
cm = confusion_matrix(Y_Test, Y_Pred)

In [188]:
print(confusion_matrix(Y_Test, Y_Pred))

[[13  0  0]
 [ 0 16  0]
 [ 0  5  4]]


In [189]:
print(classification_report(Y_Test, Y_Pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        13
           1       0.76      1.00      0.86        16
           2       1.00      0.44      0.62         9

    accuracy                           0.87        38
   macro avg       0.92      0.81      0.83        38
weighted avg       0.90      0.87      0.85        38

